pruntime/enclave/src/contracts/diem.rs

```diff
@@ -0,0 +1,281 @@
+    pub sequence_number: u64,
+    pub sent_events_key: String,
+    pub received_events_key: String,
+    pub balances: Vec<Amount>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct TransactionWithProof {
+    transaction_bytes: Vec<u8>,
+
+    epoch_change_proof: EpochChangeProof,
+    ledger_info_with_signatures: LedgerInfoWithSignatures,
+
+    ledger_info_to_transaction_info_proof: TransactionAccumulatorProof,
+    transaction_info: TransactionInfo,
+    transaction_info_to_account_proof: SparseMerkleProof,
+    account_state_blob: AccountStateBlob,
+
+    version: u64,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum Error {
+    NotAuthorized,
+    Other(String),
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum Command {
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum Request {
+    AccountData { account_data_b64: String },
+    VerifyTransaction { account_address: String, transaction_with_proof_b64: String }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum Response {
+    AccountData { size: u32 },
+    VerifyTransaction { total: u32, verified: bool },
+    Error(Error)
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Diem {
+    accounts: Vec<AccountInfo>,
+    transactions: BTreeMap<String, Vec<Transaction>>, //address => Transaction
+    verified: BTreeMap<String, bool>, //Hash => Bool
+    seq_number: BTreeMap<String, u64>, //Address => seq
+}
+
+impl Diem {
+    pub fn new() -> Self {
+        Diem {
+            accounts: Vec::new(),
+            transactions: BTreeMap::<String, Vec<Transaction>>::new(),
+            verified: BTreeMap::<String, bool>::new(),
+            seq_number: BTreeMap::<String, u64>::new(),
+        }
+    }
+}
+
+impl contracts::Contract<Command, Request, Response> for Diem {
+    fn id(&self) -> contracts::ContractId { contracts::DIEM }
+
+    fn handle_command(&mut self, _origin: &chain::AccountId, _txref: &TxRef, _cmd: Command) -> TransactionStatus {
+        TransactionStatus::Ok
+    }
+
+    fn handle_query(&mut self, origin: Option<&chain::AccountId>, req: Request) -> Response {
+        let inner = || -> Result<Response, Error> {
+            match req {
+                Request::AccountData { account_data_b64} => {
+                    println!("account_data_b64: {:?}", account_data_b64);
+                    let account_data = base64::decode(&account_data_b64).unwrap();
+                    let account_info: AccountInfo = bcs::from_bytes(&account_data).unwrap();
+                    println!("account_info:{:?}", account_info);
+                    let exist = self.accounts.iter().any(|x| x.address == account_info.address);
+                    if !exist {
+                        self.accounts.push(account_info);
+                    }
+
+                    Ok(Response::AccountData { size: self.accounts.len() as u32 })
+                }
+
+                Request::VerifyTransaction {account_address, transaction_with_proof_b64} => {
```

Comment:
Let's move this to another function, and maybe break it to a few functions? It's tooo long

Reply:
Done